File: PokedexReactASP.Domain/Enums/PokemonEnums.cs
namespace PokedexReactASP.Domain.Enums
{
    /// <summary>
    /// Pokemon Nature - affects stats by +10%/-10%
    /// Reference: https://bulbapedia.bulbagarden.net/wiki/Nature
    /// </summary>
    public enum Nature
    {
        // Neutral (no effect)
        Hardy = 0,      // Atk/Atk
        Docile = 1,     // Def/Def
        Serious = 2,    // Spd/Spd
        Bashful = 3,    // SpAtk/SpAtk
        Quirky = 4,     // SpDef/SpDef

        // Attack+
        Lonely = 5,     // +Atk, -Def
        Brave = 6,      // +Atk, -Spd
        Adamant = 7,    // +Atk, -SpAtk
        Naughty = 8,    // +Atk, -SpDef

        // Defense+
        Bold = 9,       // +Def, -Atk
        Relaxed = 10,   // +Def, -Spd
        Impish = 11,    // +Def, -SpAtk
        Lax = 12,       // +Def, -SpDef

        // Speed+
        Timid = 13,     // +Spd, -Atk
        Hasty = 14,     // +Spd, -Def
        Jolly = 15,     // +Spd, -SpAtk
        Naive = 16,     // +Spd, -SpDef

        // Sp. Attack+
        Modest = 17,    // +SpAtk, -Atk
        Mild = 18,      // +SpAtk, -Def
        Quiet = 19,     // +SpAtk, -Spd
        Rash = 20,      // +SpAtk, -SpDef

        // Sp. Defense+
        Calm = 21,      // +SpDef, -Atk
        Gentle = 22,    // +SpDef, -Def
        Sassy = 23,     // +SpDef, -Spd
        Careful = 24    // +SpDef, -SpAtk
    }

    /// <summary>
    /// Pokeball types with different catch rate multipliers
    /// </summary>
    public enum PokeballType
    {
        Pokeball = 0,       // 1.0x
        GreatBall = 1,      // 1.5x
        UltraBall = 2,      // 2.0x
        MasterBall = 3,     // Guaranteed catch
        
        // Special balls
        QuickBall = 10,     // 5.0x first turn, 1.0x after
        TimerBall = 11,     // Up to 4.0x based on turns
        DuskBall = 12,      // 3.0x at night/cave
        NetBall = 13,       // 3.5x for Water/Bug types
        DiveBall = 14,      // 3.5x underwater
        NestBall = 15,      // Better for low level
        RepeatBall = 16,    // 3.5x if already caught species
        LuxuryBall = 17,    // 1.0x but +1 friendship per catch
        PremierBall = 18,   // 1.0x (cosmetic)
        HealBall = 19,      // 1.0x but heals on catch
        
        // Rare balls
        SafariBall = 30,
        SportBall = 31,
        DreamBall = 32,
        BeastBall = 33      // 5.0x for Ultra Beasts, 0.1x otherwise
    }

    /// <summary>
    /// Pokemon overall rating based on IVs, Nature, and potential
    /// </summary>
    public enum PokemonRank
    {
        D = 0,      // < 50% potential
        C = 1,      // 50-65% potential
        B = 2,      // 65-80% potential
        A = 3,      // 80-90% potential
        S = 4,      // 90-95% potential
        SS = 5      // 95%+ potential (near perfect)
    }

    /// <summary>
    /// Result of a catch attempt
    /// </summary>
    public enum CatchAttemptResult
    {
        Success = 0,        // Pokemon caught!
        Escaped = 1,        // Ball shook but escaped
        Fled = 2,           // Pokemon ran away (rare)
        BallBroke = 3,      // Ball broke immediately
        AlreadyOwned = 4,   // For special event Pokemon
        InventoryFull = 5   // PC box full
    }

    /// <summary>
    /// Pokemon gender
    /// </summary>
    public enum PokemonGender
    {
        Male = 0,
        Female = 1,
        Genderless = 2
    }

    /// <summary>
    /// Time of day - affects certain catch rates and encounters
    /// </summary>
    public enum TimeOfDay
    {
        Morning = 0,    // 6:00 - 10:00
        Day = 1,        // 10:00 - 17:00
        Evening = 2,    // 17:00 - 20:00
        Night = 3       // 20:00 - 6:00
    }

    /// <summary>
    /// Location type - affects certain catch rates
    /// </summary>
    public enum LocationType
    {
        Grassland = 0,
        Forest = 1,
        Cave = 2,
        Water = 3,
        Mountain = 4,
        Desert = 5,
        Snow = 6,
        Urban = 7,
        Safari = 8
    }
}

File: pokedexreactasp.client/src/types/pokemon.enums.ts
/**
 * Pokemon Nature - affects stats by +10%/-10%
 */
export enum Nature {
  Hardy = 0,
  Docile = 1,
  Serious = 2,
  Bashful = 3,
  Quirky = 4,
  Lonely = 5,
  Brave = 6,
  Adamant = 7,
  Naughty = 8,
  Bold = 9,
  Relaxed = 10,
  Impish = 11,
  Lax = 12,
  Timid = 13,
  Hasty = 14,
  Jolly = 15,
  Naive = 16,
  Modest = 17,
  Mild = 18,
  Quiet = 19,
  Rash = 20,
  Calm = 21,
  Gentle = 22,
  Sassy = 23,
  Careful = 24,
}

/**
 * Pokemon Gender
 */
export enum PokemonGender {
  Male = 0,
  Female = 1,
  Genderless = 2,
}

/**
 * Pokemon Rank based on IVs
 */
export enum PokemonRank {
  D = 0,
  C = 1,
  B = 2,
  A = 3,
  S = 4,
  SS = 5,
}

/**
 * Pokeball types
 */
export enum PokeballType {
  Pokeball = 0,
  GreatBall = 1,
  UltraBall = 2,
  MasterBall = 3,
  QuickBall = 10,
  TimerBall = 11,
  DuskBall = 12,
  NetBall = 13,
  DiveBall = 14,
  NestBall = 15,
  RepeatBall = 16,
  LuxuryBall = 17,
  PremierBall = 18,
  HealBall = 19,
  SafariBall = 30,
  SportBall = 31,
  DreamBall = 32,
  BeastBall = 33,
}

/**
 * Get nature display name
 */
export function getNatureDisplay(nature: Nature): string {
  const natureEffects: Record<Nature, string> = {
    [Nature.Hardy]: "Hardy (Neutral)",
    [Nature.Docile]: "Docile (Neutral)",
    [Nature.Serious]: "Serious (Neutral)",
    [Nature.Bashful]: "Bashful (Neutral)",
    [Nature.Quirky]: "Quirky (Neutral)",
    [Nature.Lonely]: "Lonely (+Atk, -Def)",
    [Nature.Brave]: "Brave (+Atk, -Spd)",
    [Nature.Adamant]: "Adamant (+Atk, -SpA)",
    [Nature.Naughty]: "Naughty (+Atk, -SpD)",
    [Nature.Bold]: "Bold (+Def, -Atk)",
    [Nature.Relaxed]: "Relaxed (+Def, -Spd)",
    [Nature.Impish]: "Impish (+Def, -SpA)",
    [Nature.Lax]: "Lax (+Def, -SpD)",
    [Nature.Timid]: "Timid (+Spd, -Atk)",
    [Nature.Hasty]: "Hasty (+Spd, -Def)",
    [Nature.Jolly]: "Jolly (+Spd, -SpA)",
    [Nature.Naive]: "Naive (+Spd, -SpD)",
    [Nature.Modest]: "Modest (+SpA, -Atk)",
    [Nature.Mild]: "Mild (+SpA, -Def)",
    [Nature.Quiet]: "Quiet (+SpA, -Spd)",
    [Nature.Rash]: "Rash (+SpA, -SpD)",
    [Nature.Calm]: "Calm (+SpD, -Atk)",
    [Nature.Gentle]: "Gentle (+SpD, -Def)",
    [Nature.Sassy]: "Sassy (+SpD, -Spd)",
    [Nature.Careful]: "Careful (+SpD, -SpA)",
  };
  return natureEffects[nature] || "Unknown";
}

/**
 * Get gender symbol
 */
export function getGenderSymbol(gender: PokemonGender): string {
  switch (gender) {
    case PokemonGender.Male:
      return "♂";
    case PokemonGender.Female:
      return "♀";
    default:
      return "⚲";
  }
}

/**
 * Get rank display
 */
export function getRankDisplay(rank: PokemonRank): string {
  switch (rank) {
    case PokemonRank.SS:
      return "✨ SS Rank! ✨";
    case PokemonRank.S:
      return "⭐ S Rank!";
    case PokemonRank.A:
      return "A Rank";
    case PokemonRank.B:
      return "B Rank";
    case PokemonRank.C:
      return "C Rank";
    default:
      return "D Rank";
  }
}

/**
 * Catch attempt result - matches backend CatchAttemptResult enum
 */
export enum CatchAttemptResult {
  Success = 0,
  Escaped = 1,
  Fled = 2,
  BallBroke = 3,
  AlreadyOwned = 4,
  InventoryFull = 5,
}

/**
 * Get pokeball name
 */
export function getPokeballName(ball: PokeballType): string {
  const names: Record<PokeballType, string> = {
    [PokeballType.Pokeball]: "Poké Ball",
    [PokeballType.GreatBall]: "Great Ball",
    [PokeballType.UltraBall]: "Ultra Ball",
    [PokeballType.MasterBall]: "Master Ball",
    [PokeballType.QuickBall]: "Quick Ball",
    [PokeballType.TimerBall]: "Timer Ball",
    [PokeballType.DuskBall]: "Dusk Ball",
    [PokeballType.NetBall]: "Net Ball",
    [PokeballType.DiveBall]: "Dive Ball",
    [PokeballType.NestBall]: "Nest Ball",
    [PokeballType.RepeatBall]: "Repeat Ball",
    [PokeballType.LuxuryBall]: "Luxury Ball",
    [PokeballType.PremierBall]: "Premier Ball",
    [PokeballType.HealBall]: "Heal Ball",
    [PokeballType.SafariBall]: "Safari Ball",
    [PokeballType.SportBall]: "Sport Ball",
    [PokeballType.DreamBall]: "Dream Ball",
    [PokeballType.BeastBall]: "Beast Ball",
  };
  return names[ball] || "Poké Ball";
}

/**
 * Wild Spawn Rarity
 */
export enum WildSpawnRarity {
  Common = 1,
  Uncommon = 2,
  Rare = 3,
  Epic = 4,
  Legendary = 5,
}

/**
 * TCG Card Rarity Tier
 */
export enum TcgCardRarityTier {
  Common = 1,
  Uncommon = 2,
  Rare = 3,
  HoloRare = 4,
  UltraRare = 5,
  SecretRare = 6,
  Promo = 7,
  Unknown = 99,
}

/**
 * Card Reward Source
 */
export enum CardRewardSource {
  WildAreaCatch = 1,
  DailyReward = 2,
  EventReward = 3,
  AchievementReward = 4,
}

/**
 * Pokemon Location
 */
export enum PokemonLocation {
  Party = 1,
  Box = 2,
}

/**
 * Time of Day
 */
export enum TimeOfDay {
  Morning = 0,
  Day = 1,
  Evening = 2,
  Night = 3,
}

/**
 * Location Type
 */
export enum LocationType {
  Grassland = 0,
  Forest = 1,
  Cave = 2,
  Water = 3,
  Mountain = 4,
  Desert = 5,
  Snow = 6,
  Urban = 7,
  Safari = 8,
}

export function getWildSpawnRarityDisplay(rarity: WildSpawnRarity): string {
  switch (rarity) {
    case WildSpawnRarity.Common:
      return "Common";
    case WildSpawnRarity.Uncommon:
      return "Uncommon";
    case WildSpawnRarity.Rare:
      return "Rare";
    case WildSpawnRarity.Epic:
      return "Epic";
    case WildSpawnRarity.Legendary:
      return "Legendary";
    default:
      return "Unknown";
  }
}

export function getTcgCardRarityTierDisplay(tier: TcgCardRarityTier): string {
  switch (tier) {
    case TcgCardRarityTier.Common:
      return "Common";
    case TcgCardRarityTier.Uncommon:
      return "Uncommon";
    case TcgCardRarityTier.Rare:
      return "Rare";
    case TcgCardRarityTier.HoloRare:
      return "Holo Rare";
    case TcgCardRarityTier.UltraRare:
      return "Ultra Rare";
    case TcgCardRarityTier.SecretRare:
      return "Secret Rare";
    case TcgCardRarityTier.Promo:
      return "Promo";
    default:
      return "Unknown";
  }
}
